File: app/src/main/java/com/smarthome/intercom/call/CallController.kt
package com.smarthome.intercom.call

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.media.AudioManager
import android.util.Log
import android.view.Surface
import androidx.core.content.ContextCompat
import com.smarthome.intercom.config.DeviceConfig
import com.smarthome.intercom.media.AudioPipeline
import com.smarthome.intercom.media.VideoPipeline
import com.smarthome.intercom.net.CallConnection
import com.smarthome.intercom.net.CallServer
import com.smarthome.intercom.net.DiscoveryResponder
import com.smarthome.intercom.net.NetworkBinder
import com.smarthome.intercom.protocol.Channel
import com.smarthome.intercom.protocol.Commands
import com.smarthome.intercom.protocol.Frame
import com.smarthome.intercom.protocol.InboundCommand
import com.smarthome.intercom.protocol.ScreenInfo
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.net.Socket

/**
 * The single source of truth for the call. Owns discovery, the TCP call server,
 * the active connection, and the media pipelines; maps protocol events to the
 * [CallUiState] the UI observes, and turns user intents into protocol commands.
 *
 * State-changing transitions are confined to [stateLock] so socket-thread events
 * (frames arriving) and main-thread intents (Answer/End) can't race. Per-frame
 * media forwarding reads the volatile [phase] without locking to stay off the
 * hot path.
 */
class CallController(
    private val context: Context,
    private val deviceConfig: DeviceConfig,
) {
    private val scope = CoroutineScope(SupervisorJob() + kotlinx.coroutines.Dispatchers.Default)
    private val stateLock = Any()

    private val binder = NetworkBinder(context)
    private val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager

    private val discovery = DiscoveryResponder(scope, binder, ::buildScreenInfo)
    private val server = CallServer(
        scope = scope,
        bindAddressProvider = { binder.lanIpv4Address() },
        onAccepted = ::onSocketAccepted,
    )

    private val video = VideoPipeline(onError = ::onVideoError)
    private var audio: AudioPipeline? = null
    private var connection: CallConnection? = null
    private var surface: Surface? = null
    private var videoRunning = false
    private var transientJob: Job? = null

    @Volatile private var phase: CallPhase = CallPhase.IDLE

    private val _uiState = MutableStateFlow(CallUiState())
    val uiState: StateFlow<CallUiState> = _uiState.asStateFlow()

    // --- lifecycle ---

    fun start() {
        binder.acquireLocks()
        scope.launch {
            deviceConfig.identity.collect { id ->
                _uiState.update { it.copy(unitName = id.alias, pairedDoor = id.doorName, callerLabel = id.doorName) }
            }
        }
        refreshConnectivity()
        discovery.start()
        server.start()
    }

    fun shutdown() {
        teardownCall(showEnded = false)
        discovery.stop()
        server.stop()
        binder.releaseLocks()
    }

    fun refreshConnectivity() {
        val onWifi = binder.isOnLan()
        Log.d(TAG, "Connectivity: onLan=$onWifi lanIp=${binder.lanIpv4Address()?.hostAddress} networks=${binder.describeNetworks()}")
        _uiState.update { it.copy(onWifi = onWifi) }
    }

    // --- inbound: connections & frames ---

    private fun onSocketAccepted(socket: Socket) {
        synchronized(stateLock) {
            if (connection != null) {
                // If the previous connection is stale (socket closed but not cleaned up),
                // tear it down and accept the new one instead of rejecting as busy.
                val prevSocket = try { connection?.remoteAddress } catch (_: Exception) { null }
                Log.d(TAG, "Door connected from ${socket.inetAddress?.hostAddress} — previous connection exists (remote=$prevSocket, phase=$phase)")
                if (phase == CallPhase.RINGING || phase == CallPhase.IDLE) {
                    Log.d(TAG, "  Tearing down stale connection to accept new one")
                    teardownCall(showEnded = false)
                } else {
                    Log.d(TAG, "  Rejected (busy: active call in phase=$phase)")
                    rejectBusy(socket)
                    return
                }
            }
            Log.d(TAG, "Door connected from ${socket.inetAddress?.hostAddress}:${socket.port} -> opening call channel")
            val conn = CallConnection(
                socket = socket,
                scope = scope,
                onFrame = ::onFrame,
                onClosed = ::onConnectionClosed,
            )
            connection = conn
            conn.start()
        }
    }

    private fun rejectBusy(socket: Socket) {
        runCatching {
            socket.getOutputStream().apply {
                write(Commands.deviceBusy())
                flush()
            }
            socket.close()
        }
    }

    private fun onFrame(channel: Channel, payload: ByteArray) {
        when (channel) {
            Channel.CONTROL -> handleControl(payload)
            Channel.VIDEO -> if (phase != CallPhase.IDLE) {
                // Feed only — never start the decoder from this socket-thread path.
                // Attach/detach happens on the locked surface+phase transitions; doing
                // it here would race the main thread into two decoders on one Surface.
                video.submit(payload)
                if (!_uiState.value.hasVideoFrames) {
                    _uiState.update { it.copy(hasVideoFrames = true) }
                }
            }
            Channel.AUDIO -> if (phase == CallPhase.CONNECTED) audio?.playDownlink(payload)
        }
    }

    private fun handleControl(payload: ByteArray) {
        // Log the raw control JSON so a capture shows exactly what the door sent
        // (command name / casing / shape) when classification doesn't match.
        Log.d(TAG, "CONTROL rx: ${payload.decodeToString()}")
        val msg = Commands.parse(payload)
        if (msg == null) {
            Log.d(TAG, "  Unparseable control payload")
            return
        }
        val command = msg.classify()
        Log.d(TAG, "  parsed name='${msg.name}' -> classified=$command")
        when (command) {
            InboundCommand.CALL -> onCall()
            InboundCommand.GET_CALL_INFO -> onGetCallInfo()
            InboundCommand.HANG_UP -> onRemoteHangUp()
            InboundCommand.UNKNOWN -> Log.d(TAG, "  Unknown command: ${msg.name}")
        }
    }

    private fun onCall() = synchronized(stateLock) {
        Log.d(TAG, "onCall() invoked, current phase=$phase, connection=${connection != null}")
        if (phase != CallPhase.IDLE) {
            Log.w(TAG, "Call received but ignored; already in phase=$phase — if stuck, restart the app")
            return // already busy with a call
        }
        Log.d(TAG, "Incoming Call -> RINGING")
        phase = CallPhase.RINGING
        val label = deviceConfig.identity.value.doorName
        _uiState.update {
            it.copy(
                phase = CallPhase.RINGING,
                callerLabel = label,
                videoAvailable = true,
                hasVideoFrames = false,
                muted = false,
                transientMessage = null,
            )
        }
        ensureVideoStarted()
    }

    private fun onGetCallInfo() = synchronized(stateLock) {
        // Re-confirm our answer if the door re-queries mid-call.
        if (phase == CallPhase.CONNECTED) {
            val conn = connection ?: return
            Commands.answerFrames().forEach { conn.enqueue(it) }
        }
    }

    private fun onRemoteHangUp() = synchronized(stateLock) {
        if (phase == CallPhase.IDLE) return
        teardownCall(showEnded = true)
    }

    private fun onConnectionClosed() = synchronized(stateLock) {
        connection = null
        if (phase != CallPhase.IDLE) teardownCall(showEnded = true) else stopVideo()
    }

    // --- user intents ---

    fun answer() = synchronized(stateLock) {
        if (phase != CallPhase.RINGING) return
        val conn = connection ?: run {
            teardownCall(showEnded = true)
            return
        }
        phase = CallPhase.CONNECTING
        _uiState.update { it.copy(phase = CallPhase.CONNECTING) }

        // Answer handshake: the three Answer variants, in order. The reference
        // client sends no StartTalk on accept, so we don't either.
        Commands.answerFrames().forEach { conn.enqueue(it) }

        startAudio()

        phase = CallPhase.CONNECTED
        _uiState.update { it.copy(phase = CallPhase.CONNECTED) }
    }

    fun decline() = synchronized(stateLock) {
        if (phase == CallPhase.IDLE) return
        connection?.enqueue(Commands.hangUp())
        teardownCall(showEnded = false)
    }

    /** End an active call (same protocol as decline; distinct intent for clarity). */
    fun endCall() = decline()

    fun unlock() = synchronized(stateLock) {
        // SEC-2: only from a connected, on-screen call.
        if (phase != CallPhase.CONNECTED) return
        connection?.enqueue(Commands.openDoor())
        showTransient(MESSAGE_UNLOCKED)
    }

    fun setMuted(muted: Boolean) = synchronized(stateLock) {
        audio?.setMuted(muted)
        _uiState.update { it.copy(muted = muted) }
    }

    /**
     * Re-arm capture when the mic is granted *during* a call. The pipeline was
     * started without an uplink (permission was denied at answer time), so we
     * restart it now that recording is allowed.
     */
    fun onMicPermissionResult(granted: Boolean) = synchronized(stateLock) {
        if (!granted || phase != CallPhase.CONNECTED) return
        if (_uiState.value.micAvailable) return // already capturing
        audio?.stop()
        audio = null
        startAudio()
    }

    fun setVideoSurface(newSurface: Surface) = synchronized(stateLock) {
        if (newSurface === surface) return
        surface = newSurface
        if (phase != CallPhase.IDLE) {
            // A new SurfaceView (ring screen -> in-call screen) replaced the old one:
            // detach the decoder bound to the previous surface, then re-attach.
            stopVideo()
            ensureVideoStarted()
        }
    }

    /**
     * The given surface's [SurfaceView] was destroyed. Only tear down if it's still
     * the active surface: when screens swap, the outgoing view's destroy callback can
     * arrive *after* the incoming view's create, and must not kill the new decoder.
     */
    fun clearVideoSurface(oldSurface: Surface) = synchronized(stateLock) {
        if (oldSurface !== surface) return
        surface = null
        stopVideo()
    }

    fun clearTransientMessage() {
        _uiState.update { it.copy(transientMessage = null) }
    }

    // --- media + teardown helpers (call holding stateLock, except submit paths) ---

    private fun startAudio() {
        val micGranted = hasRecordPermission()
        val pipeline = AudioPipeline(
            audioManager = audioManager,
            scope = scope,
            onAudioEncoded = { alaw -> connection?.enqueue(Frame.encode(Channel.AUDIO, alaw)) },
            onMicUnavailable = { _uiState.update { it.copy(micAvailable = false) } },
        )
        pipeline.start(captureEnabled = micGranted)
        pipeline.setMuted(_uiState.value.muted)
        audio = pipeline
        _uiState.update { it.copy(micAvailable = micGranted) }
    }

    private fun ensureVideoStarted() {
        val s = surface ?: return
        if (!videoRunning) {
            video.start(s)
            videoRunning = true
        }
    }

    private fun stopVideo() {
        if (videoRunning) {
            video.stop()
            videoRunning = false
        }
    }

    private fun onVideoError() = synchronized(stateLock) {
        videoRunning = false
        _uiState.update { it.copy(videoAvailable = false) }
    }

    private fun teardownCall(showEnded: Boolean) {
        phase = CallPhase.IDLE
        audio?.stop()
        audio = null
        stopVideo()
        val conn = connection
        connection = null
        conn?.close()
        _uiState.update {
            it.copy(
                phase = CallPhase.IDLE,
                muted = false,
                hasVideoFrames = false,
                videoAvailable = true,
                micAvailable = true,
            )
        }
        if (showEnded) showTransient(MESSAGE_ENDED)
    }

    private fun showTransient(message: String) {
        _uiState.update { it.copy(transientMessage = message) }
        transientJob?.cancel()
        transientJob = scope.launch {
            delay(TRANSIENT_MS)
            _uiState.update { if (it.transientMessage == message) it.copy(transientMessage = null) else it }
        }
    }

    private fun buildScreenInfo(): ScreenInfo {
        val id = deviceConfig.identity.value
        return ScreenInfo(
            alias = id.alias,
            serial = id.serial,
            dstAddr = id.dstAddr,
        )
    }

    private fun hasRecordPermission(): Boolean =
        ContextCompat.checkSelfPermission(context, Manifest.permission.RECORD_AUDIO) ==
            PackageManager.PERMISSION_GRANTED

    private companion object {
        const val TAG = "CallController"
        const val TRANSIENT_MS = 2500L
        const val MESSAGE_UNLOCKED = "Door unlocked"
        const val MESSAGE_ENDED = "Call ended"
    }
}

File: packages/intercom/lib/src/call/call_controller.dart
import 'dart:async';
import 'dart:io';

import 'package:connectivity_plus/connectivity_plus.dart';
import 'package:flutter/foundation.dart';
import 'package:permission_handler/permission_handler.dart';

import '../background/android_foreground_service.dart';
import '../config/device_config.dart';
import '../intercom_mode.dart';
import '../media/audio_pipeline.dart';
import '../media/video_decoder.dart';
import '../net/call_connection.dart';
import '../net/call_server.dart';
import '../net/discovery_responder.dart';
import '../net/local_ip_address.dart';
import '../notifications/intercom_notification_handler.dart';
import '../notifications/notification_actions.dart';
import '../protocol/commands.dart';
import '../protocol/discovery.dart';
import '../protocol/frame.dart';
import '../ringer/ringer.dart';
import 'call_phase.dart';
import 'call_ui_state.dart';

final class CallController extends ChangeNotifier {
  CallController({
    required this.mode,
    required this.deviceConfig,
    Ringer? ringer,
    VideoDecoder? video,
    AudioPipeline? audio,
  })  : _ringer = ringer ?? Ringer(),
        _video = video ?? VideoDecoder(),
        _audio = audio ?? AudioPipeline() {
    final id = deviceConfig.identity;
    _state = _state.copyWith(
        unitName: id.alias, pairedDoor: id.doorName, callerLabel: id.doorName);
    _server = CallServer(
      onAccepted: _onSocketAccepted,
      onError: _onServerError,
    );
    _discovery = DiscoveryResponder(screenInfoProvider: _buildScreenInfo);
    _notifications = IntercomNotificationHandler(onTap: _onNotificationTap);
  }

  final IntercomMode mode;
  final DeviceConfig deviceConfig;
  final Ringer _ringer;
  final VideoDecoder _video;
  final AudioPipeline _audio;
  late final IntercomNotificationHandler _notifications;
  late final CallServer _server;
  late final DiscoveryResponder _discovery;
  CallConnection? _connection;
  StreamSubscription? _audioSub;
  Timer? _transientTimer;
  CallUiState _state = const CallUiState();

  CallUiState get state => _state;
  int? get videoTextureId => _video.textureId;

  void refreshIdentity() {
    final id = deviceConfig.identity;
    _setState(_state.copyWith(
      unitName: id.alias,
      pairedDoor: id.doorName,
      callerLabel: id.doorName,
    ));
  }

  Future<void> start() async {
    await _startupStep('notifications.initialize', _notifications.initialize);
    await _startupStep('permission.notification.request',
        () => Permission.notification.request());
    await _startupStep('connectivity.refresh', refreshConnectivity);
    if (mode == IntercomMode.panel) {
      final discoveryStarted =
          await _startupStep('discovery.start', _discovery.start);
      final serverStarted =
          await _startupStep('call_server.start', _server.start);
      await _startupStep('foreground.startPanelService',
          AndroidForegroundService.startPanelService);
      _setState(_state.copyWith(
        discoveryListening: discoveryStarted,
        tcpServerListening: serverStarted,
      ));
      await _startupStep(
          'foreground.takePendingIncomingCall', checkPendingBackgroundCall);
    }
  }

  Future<T> _startupStep<T>(
    String name,
    Future<T> Function() action,
  ) async {
    const timeout = Duration(seconds: 4);
    debugPrint('Intercom startup step started: $name');
    final stopwatch = Stopwatch()..start();
    try {
      final result = await action().timeout(timeout);
      debugPrint(
          'Intercom startup step finished: $name (${stopwatch.elapsedMilliseconds}ms)');
      return result;
    } on TimeoutException catch (error, stackTrace) {
      final message =
          'Intercom startup step timed out: $name after ${timeout.inSeconds}s';
      debugPrint(message);
      debugPrintStack(stackTrace: stackTrace);
      throw TimeoutException(message, error.duration);
    } catch (error, stackTrace) {
      debugPrint('Intercom startup step failed: $name -> $error');
      debugPrintStack(stackTrace: stackTrace);
      Error.throwWithStackTrace(Exception('$name failed: $error'), stackTrace);
    }
  }

  Future<void> checkPendingBackgroundCall() async {
    if (mode != IntercomMode.panel || _state.phase != CallPhase.idle) return;
    final hasPendingCall =
        await AndroidForegroundService.takePendingIncomingCall();
    if (hasPendingCall) {
      final id = deviceConfig.identity;
      _setState(_state.copyWith(
        phase: CallPhase.ringing,
        callerLabel: id.doorName,
        videoAvailable: true,
        hasVideoFrames: false,
        muted: false,
        transientMessage: null,
      ));
    }
  }

  Future<void> shutdown() async {
    await _teardownCall(showEnded: false);
    await _discovery.stop();
    await _server.stop();
    if (mode == IntercomMode.panel) {
      await AndroidForegroundService.stopPanelService();
    }
    _setState(_state.copyWith(
      discoveryListening: false,
      tcpServerListening: false,
    ));
  }

  Future<void> refreshConnectivity() async {
    final result = await Connectivity().checkConnectivity();
    final localIpAddress = await resolveLocalIpv4Address();
    _setState(_state.copyWith(
      onWifi: result.contains(ConnectivityResult.wifi),
      localIpAddress: localIpAddress,
    ));
  }

  Future<void> connectToDoor(String host,
      {int port = CallServer.defaultPort}) async {
    final socket =
        await Socket.connect(host, port, timeout: const Duration(seconds: 5));
    _onSocketAccepted(socket);
  }

  Future<void> simulateIncomingCall() async {
    if (_state.phase != CallPhase.idle) return;
    final id = deviceConfig.identity;
    _setState(_state.copyWith(
      phase: CallPhase.ringing,
      callerLabel: id.doorName,
      videoAvailable: true,
      hasVideoFrames: false,
      muted: false,
      transientMessage: null,
    ));
    await _notifications.showIncomingCall(doorName: id.doorName);
    await _video.start();
    await _ringer.start();
  }

  Future<void> answer() async {
    if (_state.phase != CallPhase.ringing) return;
    final conn = _connection;
    if (conn == null) {
      await _teardownCall(showEnded: true);
      return;
    }
    await _notifications.dismissIncomingCall();
    await _ringer.stop();
    _setState(_state.copyWith(phase: CallPhase.connecting));
    for (final frame in Commands.answerFrames()) {
      conn.enqueue(frame);
    }
    await _startAudio();
    _setState(_state.copyWith(phase: CallPhase.connected));
  }

  Future<void> decline() async {
    if (_state.phase == CallPhase.idle) return;
    await _notifications.dismissIncomingCall();
    _connection?.enqueue(Commands.hangUp());
    await _teardownCall(showEnded: false);
  }

  Future<void> endCall() => decline();

  void unlock() {
    if (_state.phase != CallPhase.connected) return;
    _connection?.enqueue(Commands.openDoor());
    _showTransient('Door unlocked');
  }

  Future<void> setMuted(bool muted) async {
    await _audio.setMuted(muted);
    _setState(_state.copyWith(muted: muted));
  }

  void _onSocketAccepted(Socket socket) {
    if (_connection != null &&
        _state.phase != CallPhase.idle &&
        _state.phase != CallPhase.ringing) {
      socket.add(Commands.deviceBusy());
      socket.destroy();
      return;
    }
    _connection?.close();
    final conn = CallConnection(
      socket: socket,
      onFrame: _onFrame,
      onClosed: () => _teardownCall(showEnded: true),
    );
    _connection = conn;
    conn.start();
  }

  void _onServerError(Object error) {
    if (error is SocketException && error.osError?.errorCode == 98) {
      _showTransient('Port 8189 is already in use');
      return;
    }
    _showTransient('Intercom server unavailable');
  }

  Future<void> _onFrame(Channel channel, Uint8List payload) async {
    switch (channel) {
      case Channel.control:
        await _handleControl(payload);
      case Channel.video:
        if (_state.phase != CallPhase.idle) {
          await _video.submit(payload);
          if (!_state.hasVideoFrames)
            _setState(_state.copyWith(hasVideoFrames: true));
        }
      case Channel.audio:
        if (_state.phase == CallPhase.connected)
          await _audio.playDownlink(payload);
    }
  }

  Future<void> _handleControl(Uint8List payload) async {
    final message = Commands.parse(payload);
    switch (message?.classify() ?? InboundCommand.unknown) {
      case InboundCommand.call:
        if (_state.phase != CallPhase.idle) return;
        final id = deviceConfig.identity;
        _setState(_state.copyWith(
          phase: CallPhase.ringing,
          callerLabel: id.doorName,
          videoAvailable: true,
          hasVideoFrames: false,
          muted: false,
          transientMessage: null,
        ));
        await _notifications.showIncomingCall(doorName: id.doorName);
        await _video.start();
        await _ringer.start();
      case InboundCommand.getCallInfo:
        if (_state.phase == CallPhase.connected) {
          for (final frame in Commands.answerFrames()) {
            _connection?.enqueue(frame);
          }
        }
      case InboundCommand.hangUp:
        await _teardownCall(showEnded: true);
      case InboundCommand.unknown:
        return;
    }
  }

  Future<void> _startAudio() async {
    var micStatus = await Permission.microphone.status;
    if (!micStatus.isGranted) {
      micStatus = await Permission.microphone.request();
    }
    final micGranted = micStatus.isGranted;
    await _audio.start(captureEnabled: micGranted);
    _audioSub?.cancel();
    _audioSub = _audio.uplink.listen((alaw) {
      _connection?.enqueue(Frame.encode(Channel.audio, alaw));
    });
    _setState(_state.copyWith(micAvailable: micGranted));
  }

  Future<void> _teardownCall({required bool showEnded}) async {
    await _notifications.dismissIncomingCall();
    await _ringer.stop();
    await _audioSub?.cancel();
    _audioSub = null;
    await _audio.stop().catchError((_) {});
    await _video.stop().catchError((_) {});
    final conn = _connection;
    _connection = null;
    await conn?.close();
    _setState(_state.copyWith(
      phase: CallPhase.idle,
      muted: false,
      hasVideoFrames: false,
      videoAvailable: true,
      micAvailable: true,
    ));
    if (showEnded) _showTransient('Call ended');
  }

  ScreenInfo _buildScreenInfo() {
    final id = deviceConfig.identity;
    return ScreenInfo(
      alias: id.alias,
      serial: id.serial,
      dstAddr: id.dstAddr,
      localIp: _state.localIpAddress,
    );
  }

  void _showTransient(String message) {
    _setState(_state.copyWith(transientMessage: message));
    _transientTimer?.cancel();
    _transientTimer = Timer(const Duration(milliseconds: 2500), () {
      if (_state.transientMessage == message)
        _setState(_state.copyWith(transientMessage: null));
    });
  }

  void _setState(CallUiState state) {
    _state = state;
    notifyListeners();
  }

  Future<void> _onNotificationTap(
    IntercomNotificationAction action,
    Map<String, dynamic> data,
  ) async {
    switch (action) {
      case IntercomNotificationAction.answer:
        await answer();
      case IntercomNotificationAction.reject:
        await decline();
    }
  }

  @override
  void dispose() {
    _transientTimer?.cancel();
    unawaited(shutdown());
    super.dispose();
  }
}
